Fix MSVC UWP builds

janisozaur opened the pull request:
Hi,
As mentioned on the mailing list, I'm submitting a pull request. This already includes the fixes you suggested.

janisozaur commented:
ping @IntelOrca

janisozaur commented on lib/zip_source_win32uwp.c:
```diff
@@ -0,0 +1,141 @@
+{
+    DeleteFileW((const wchar_t *)fname);
+    return 0;
+}
```
`CreateFile2`, which was the main reason behind this file, seems to be available in non-UWP builds. See documentation (https://msdn.microsoft.com/en-us/library/windows/desktop/hh449422(v=vs.85).aspx)

janisozaur replied:
Using this would raise the minimum supported version of Windows to Windows 8.

dillof replied:
This code is only used on Universal Windwos Platform. Did that exist before Windows 8?

janisozaur replied:
Did UWP exist before Win8? No, I think it is a bit more complex. UWP seems to be what WinRT evolved into and while some APIs were introduced in Win8 already, the UWP only became a thing with Windows 10

janisozaur commented on lib/zip_source_win32handle.c:
```diff
@@ -452,6 +455,10 @@ _win32_create_temp_file(_zip_source_win32_read_file_t *ctx)
+    value = (zip_uint32_t)GetTickCount64();
+#endif
+
     for (i = 0; i < 1024 && th == INVALID_HANDLE_VALUE; i++) {
```
The documentation (https://msdn.microsoft.com/en-us/library/windows/desktop/aa363875(v=vs.85).aspx) suggest using `GetTempFileName`, but it is discussed and discarded in the comment at the top of this function, so no.

janisozaur commented on lib/zip_source_win32handle.c:
```diff
@@ -420,12 +420,15 @@ _win32_create_temp_file(_zip_source_win32_read_file_t *ctx)
+#ifndef WINRT
     DWORD len;
     BOOL success;
+    SECURITY_ATTRIBUTES sa;
```
Reading between the lines of https://docs.microsoft.com/en-us/windows/uwp/files/file-access-permissions it looks like the intended usage is for the sandboxing layer to manage that. There are severe restrictions in UWP compared to regular "desktop" applications with regards to file access.

0-wiz-0 replied:
I think the point of the code here is to duplicate the ACLs from the original file as best as possible when creating a new file. This does not have anything to do with which parts of the file system are accessible, which is what the link points to. Am I missing something?

janisozaur commented on lib/zip_random_win32.c:
```diff
@@ -48,4 +49,8 @@ zip_random(zip_uint8_t *buffer, zip_uint16_t length)
+#else
+    // Not implemented yet.
+    return false;
+#endif // WINRT
```
Would filling the buffer with [random value `4`](https://xkcd.com/221/) suffice? Or can you suggest some other alternative?

Sorry, but I'm not coming anywhere close the Win-CLR-COM-managed-something APIs with `^` and `%` that would require doing it _properly_ (see https://stackoverflow.com/questions/16770124/how-to-encrypt-and-decryt-a-const-char-in-winrt)

If you find this a deal breaker, suggest a way of doing it, find someone else to do it or close this PR. This is not a scenario I'm interested in and don't know how to do properly.

Please let me know how you want this handled.

0-wiz-0 replied:
`GenerateRandom(unsigned int length)` from the first link looks like a promising function. Is it usable in our context?

dillof commented:
According to UWP documentation at https://docs.microsoft.com/en-us/windows/uwp/security/common-cryptography-tasks the following should work:

bool
zip_random(zip_uint8_t *data, zip_uint16_t length)
{
    IBuffer buffer = CryptographicBuffer.GenerateRandom(length);
    CryptographicBuffer.CopyToByteArray(buffer, data, length);
    
    return true;
}

(It’s completely untested, since I don’t have a Windows system.)

janisozaur commented:
> According to UWP documentation at https://docs.microsoft.com/en-us/windows/uwp/security/common-cryptography-tasks the following should work:

Since when is libzip written in C#?

0-wiz-0 commented:
Does `BCryptGenRandom` https://msdn.microsoft.com/en-us/library/windows/desktop/aa375458(v=vs.85).aspx help for the crypto part?

janisozaur commented:
@0-wiz-0 thanks, it looks like CNG is actually available for UWP target and I implemented the entropy generation with this API.

Here's a link to an appveyor job producing UWP binaries (it still timeouts due to MSVC's terrible sluggishness): https://ci.appveyor.com/project/janisozaur/vcpkg/build/1.0.13

Is there anything left before this can get merged?

janisozaur commented:
I have addressed your requests:

* DACLs should now be preserved even on UWP builds
* Moved myself to `THANKS` file

The DACL bit was particularly hard because of misleading, inconsistent and buggy documentation by Microsoft.

Please find the test build here: https://ci.appveyor.com/project/janisozaur/vcpkg/build/1.0.24

janisozaur commented:
Is extracting UWP code from `zip_random_win32.c` to its own file the last thing that's needed before this can get merged?

janisozaur commented:
I have extracted the code you requested to its own file

janisozaur commented:
Is there still anything to be done here?

0-wiz-0 commented:
I don't think so.
Thank you. We'll look at it closely and merge it in the next ten days.

0-wiz-0 commented:
We did some minor cleanups and committed this as https://github.com/nih-at/libzip/commit/2db8ea70d5ec8b58c9fe1318f4572888dc00943d.
Thank you very much!

janisozaur commented:
Great, I will update vcpkg recipe once you do another release – when can I expect that to happen?
